### DISCUSSION TITLE
develop_from - descendants problem

### DISCUSSION THREAD

**Discussion by noamiz5060** [networkx/networkx#6395] [Q&A]
Hello hello!

I'm using netwrokx with cell ontology obo file, specifically the ancestors and descendants functions. The problem is it includes the develop_from relationship, which I want to avoid 
![image](https://user-images.githubusercontent.com/111743864/215234487-47c8de48-0b8f-48e9-87d0-b323b1b0b950.png)
here, for exmaple, Bm3-delta B cell doesn't have descendants, but the ancestors function includes the develop_from, in this case Bm 4 B cell. 
Is there any way I can limit it to 'normal' descendants?

**Comment by rossbar** [comment]
Typically it's best to provide information related to the NetworkX portion of the code/analysis rather than screenshots of unrelated software. I can't tell from the provided info if this is what you're asking, but I'll try to answer anyway!

Let's assume you have a cell-lineage like so:

```python
>>> G = nx.DiGraph([("Bm4_Bc", "gcBc"), ("gcBc", "Bm3-d_Bc")])
```

where the names are rough abbreviations from the example above. In this case I'm assuming the edges represent the "ancestral" relationship, i.e. Bm4_Bc is a parent of gcBc is a parent of Bm3-d_Bc.

In this case, the `descendants` function will give you what you expect: namely that Bm3-d_Bc is the (only) child of gcBc:

```python
>>> nx.descendants(G, "gcBc")
{'Bm3-d_Bc'}
```

Note that the graph in the above example is *directed*, and the direction of the edge specifies the nature of the relationship (parent -> child). `nx.descendants` *also* works for undirected graphs, in which case the edges can be thought of as bidirectional. With respect to the above example, that would mean "Bm4_Bc" would also be a valid child of "gcBc":

```python
>>> nx.descendants(G.to_undirected(), "gcBc")
{'Bm3-d_Bc', 'Bm4_Bc'}
```

If I've understood the initial post correctly, this seems like a likely culprit. Are you expecting edge direction to specify relationship? If so, double-check that you are using a directed graph.

**Reply by noamiz5060** [reply]
> Typically it's best to provide information related to the NetworkX portion of the code/analysis rather than screenshots of unrelated software. I can't tell from the provided info if this is what you're asking, but I'll try to answer anyway!
> 
> Let's assume you have a cell-lineage like so:
> 
> ```python
> >>> G = nx.DiGraph([("Bm4_Bc", "gcBc"), ("gcBc", "Bm3-d_Bc")])
> ```
> 
> where the names are rough abbreviations from the example above. In this case I'm assuming the edges represent the "ancestral" relationship, i.e. Bm4_Bc is a parent of gcBc is a parent of Bm3-d_Bc.
> 
> In this case, the `descendants` function will give you what you expect: namely that Bm3-d_Bc is the (only) child of gcBc:
> 
> ```python
> >>> nx.descendants(G, "gcBc")
> {'Bm3-d_Bc'}
> ```
> 
> Note that the graph in the above example is _directed_, and the direction of the edge specifies the nature of the relationship (parent -> child). `nx.descendants` _also_ works for undirected graphs, in which case the edges can be thought of as bidirectional. With respect to the above example, that would mean "Bm4_Bc" would also be a valid child of "gcBc":
> 
> ```python
> >>> nx.descendants(G.to_undirected(), "gcBc")
> {'Bm3-d_Bc', 'Bm4_Bc'}
> ```
> 
> If I've understood the initial post correctly, this seems like a likely culprit. Are you expecting edge direction to specify relationship? If so, double-check that you are using a directed graph.

First of all, thank you very much for replying
In my code I need to pull out all the cells who have no descendants, which one of them is Bm 3 Delta b cell and for that I used the nx.ancestors function (which returns the cells' 'children'). To get the cells, I took every cell that the function returned nothing 
![image](https://user-images.githubusercontent.com/111743864/215264953-6f658d42-931a-40ce-8ae7-dd60e59618d7.png)
So when I apply the function on Bm 3 Delta b cell, I get the Bm4 bc, which is NOT a subtype of his, but in a develops_from relationship. I'm trying to find a way which I can get the edges WITHOUT the develops_from relationships.
Hope I was a bit more clearer

**Reply by dschult** [reply]
There is a concept in graph theory called "degree" which gives the number of edges for a node. For directed graphs there is even a function `G.in_degree` and another called `G.out_degree`.  These give the number of edges that come in (or out) of a node. So some code that woud collect the nodes with no outgoing edges could look like:

```python
childless_nodes = [node for node, out_degree in G.out_degree if out_degree == 0]
```
I think you are also asking how to rule out certain kinds of edges.  The easiest way to rule them out is to not create them in the first place.   :)    You seem to be gettting the graph info from an obo file. But I don't know any function that reads an obo file and creates a networkx graph.  Perhaps you should ask the person that wrote the obo-file-reading function how to avoid creating those edges.

There are, of course, ways to ignore edges in NetworkX. But they rely on some sort of data about which edges to ignore and that depends on how you store the data in your NetworkX Graph. The graph object is very flexible and thus we can't know how you store the information for the edges.

**Reply by noamiz5060** [reply]
Thank you very much for the reply
